align/affineGap.go

```diff
@@ -1,259 +1,352 @@
+		} else if i < mColumn-1 {
+			mRowPrevious, mRowCurrent = mRowCurrent, mRowPrevious //reuse mRow variables to save memory, but only up until the second to last row
 		}
 	}
-	maxScore, route := affineTrace(mRowCurrent, mColumn, trace)
-
-	return maxScore, route
+	score_highest_i := mColumn - 1
+	score_highest_j := len(mRowCurrent[0]) - 1                                                                                            //score_highest_j was denoted with "last J" in affineGap_highMem.go
+	score_highest, _ := tripleMaxTrace(mRowCurrent[0][score_highest_j], mRowCurrent[1][score_highest_j], mRowCurrent[2][score_highest_j]) //denoted by "maxScore" in affineGap_highMem.go. The 2nd output variable is its direction (0,1,2)
+	return score_highest, score_highest_i, score_highest_j, trace_prep_i, trace_prep_j
 }
 
-func multipleAffineGapChunk(alpha []fasta.Fasta, beta []fasta.Fasta, scores [][]int64, gapOpen int64, gapExtend int64, chunkSize int64) (int64, []Cigar) {
-	var alphaSize, betaSize int64 = int64(len(alpha[0].Seq)), int64(len(beta[0].Seq))
-	if alphaSize%chunkSize != 0 {
-		log.Fatalf(fmt.Sprintf("Error: the first subalignment has a length of %d, when it should be a multiple of %d\n", alphaSize, chunkSize))
+//Step 2
+//inputs: the sequences to be aligned alpha and beta, scoring matrix for base matches, penalty for opening and extending gaps in alignment, checkerboard size for row (i) and column (j), row (i) and column (j) positions of the highest score, trace prep matrices for rows (i) and columns (j)
+//				coordinates specifying the current checkerboard in rows (k1) and columns (k2), row (i) and column (j) positions of inChecker_min_Previous which describe where Step 3 (writeCigar) stopped in the previous checkerboard that just went through Step 3 (writeCigar)
+//				trace matrix which was initialized in the main ConstGap function and passed back and forth to be recycled
+//outputs: trace matrix to be recycled, row (i), column (j) and dimension (k) positions of inChecker_max which describe where Step 2 (fillTraceback) stopped in the current checkerboard that just went through Step 2 (fillTraceback)
+func fillTraceback_affineGap(alpha []dna.Base, beta []dna.Base, scores [][]int64, gapOpen int64, gapExtend int64, checkersize_i int, checkersize_j int, score_highest_i int, score_highest_j int, trace_prep_i [][][]int64, trace_prep_j [][][]int64, k1 int, k2 int, i_inChecker_min_Previous int, j_inChecker_min_Previous int, trace [][][]ColType) ([][][]ColType, int, int, ColType) {
+	mRowCurrent := make([][]int64, 3)
+	mRowPrevious := make([][]int64, 3)
+	for k := range mRowCurrent { //k ranges through 3 numbers (0,1,2)
+		mRowCurrent[k] = make([]int64, len(beta)+1)
+		mRowPrevious[k] = make([]int64, len(beta)+1)
+		copy(mRowPrevious[k], trace_prep_i[k][k1])
+	}
+
+	//i_inChecker and j_inChecker are internal (within checkerboard) versions of i and j, e.g. for checkersize=3, i_inChecker can take on the values 0, 1, 2
```

Comment:
Honestly, I'm struggling to make sense of this note. Could you explain it more for me please?

Reply:
In a 3x3 checkerboard structure, the i and j could be 3,4,5 ("external", true i and j), but when we divide up all the i and j into checkerboards, those same coordinates would have "internal" i and j of 0,1,2. So the external i and j will keep going up (0,1,2,3,4,5,6,...) but the internal i and j will loop within the size of the checkerboard (0,1,2,0,1,2,...)